prio_processor/spark/udf.py

```diff
@@ -0,0 +1,253 @@
+import logging
+from collections import Counter
+import array
+
+import pandas as pd
+from prio import PrioContext, libprio
+from prio_processor.prio.commands import import_keys, import_public_keys, match_server
+
+logger = logging.getLogger(__name__)
+
+
+def encode(
+    batch_id: bytes,
+    n_data: int,
+    public_key_hex_internal: bytes,
+    public_key_hex_external: bytes,
+    payload: pd.Series,
+) -> pd.DataFrame:
+    with PrioContext():
+        public_key_internal, public_key_external = import_public_keys(
+            public_key_hex_internal, public_key_hex_external
+        )
+        config = libprio.PrioConfig_new(
+            n_data, public_key_internal, public_key_external, batch_id
+        )
+        results = []
+        for data in payload:
+            # trying to encode the integer into a bit array
+            a, b = libprio.PrioClient_encode(config, bytes(list(map(int, data))))
+            results.append(dict(a=a, b=b))
+        return pd.DataFrame(results)
+
+
+def verify1(
+    batch_id: bytes,
+    n_data: int,
+    server_id: str,
+    private_key_hex: bytes,
+    shared_secret: bytes,
+    public_key_hex_internal: bytes,
+    public_key_hex_external: bytes,
+    shares: pd.Series,
+) -> pd.Series:
+    # NOTE: there is an initialization call without the corresponding clear
+    # function because clearing the NSS context before the work is done causes
+    # issues when working across threads and processes. Initialization is
+    # idempotent, and it shouldn't matter if it's being called more than once.
+    # Not entirely sure why it's necessary here, but not in other grouped map
+    # functions.
+    libprio.Prio_init()
+
+    def _process(share):
+        private_key, public_key_internal, public_key_external = import_keys(
+            private_key_hex, public_key_hex_internal, public_key_hex_external
+        )
+        config = libprio.PrioConfig_new(
+            n_data, public_key_internal, public_key_external, batch_id
+        )
+        server = libprio.PrioServer_new(
+            config, match_server(server_id), private_key, shared_secret
+        )
+        verifier = libprio.PrioVerifier_new(server)
+        packet = libprio.PrioPacketVerify1_new()
+        # share is actually a bytearray, so convert it into bytes
+        libprio.PrioVerifier_set_data(verifier, bytes(share))
+        libprio.PrioPacketVerify1_set_data(packet, verifier)
+        return libprio.PrioPacketVerify1_write(packet)
+
+    results = [_process(share) for share in shares]
+    return pd.Series(results, name="payload")
+
+
+def verify2(
+    batch_id: bytes,
+    n_data: int,
+    server_id: str,
+    private_key_hex: bytes,
+    shared_secret: bytes,
+    public_key_hex_internal: bytes,
+    public_key_hex_external: bytes,
+    shares: pd.Series,
+    internal: pd.Series,
+    external: pd.Series,
+) -> pd.Series:
+    libprio.Prio_init()
+
+    def _process(share, internal, external):
```

Comment:
Quite a bit of duplication between this and `_process` in verify1. Could you at least factor out the first part of this function (up to where you create the verifier variable) into a common function?